### PR TITLE
feat(sft): generic SFT domain + qwen3(AR/CE) & sd3(diffusion/MSE) tasks

### PR DESCRIPTION
**Pull request by leviking98z-rgb**
## Summary

Adds a **generic supervised-finetuning (SFT) training domain** to UniRL — the first non-RL domain — and proves it is **model-agnostic** by shipping two validation task adapters over **already-in-tree** models: **Qwen3 (autoregressive, cross-entropy)** and **SD3.5 (diffusion, flow-matching MSE)**.

Where GRPO/ReFL do `rollout → reward → advantage → update`, SFT does the plain supervised loop `read data → compute loss → backward → update` — no rollout engine, reward, advantages, or weight sync. It mirrors the ReFL domain's shape (`BaseTrainer` + FSDP-wrapped policy), the codified template for non-GRPO domains.

**Design: the skeleton knows no model.** `SFTPolicy` resolves a **task adapter** from config (`task_target` → `get_class(...).from_config(model_config)`) and drives it; everything family-specific — which model, how to read a record, and *what the loss is* — lives in the task. That is why one skeleton serves both AR and diffusion:

| Task | Family | Loss |
|------|--------|------|
| `Qwen3SFTTask` | autoregressive | masked next-token **cross-entropy** |
| `SD3SFTTask` | diffusion | flow-matching **velocity MSE** (target `noise - x0`) |

The contract is codified in `unirl/train/sft/task.py` (`SFTTask` Protocol + `SFTTaskBase` ABC).

### What's here
- **Skeleton** (extracted from #181, cosmos-free): `unirl/trainer/sft.py`, `unirl/train/sft/{__init__,policy,data,task}.py`, `unirl/train_sft.py`.
- **Qwen3 task** (`unirl/models/qwen3/sft_task.py`): reuses the AR RL path's `_replay_aware_forward` (chunked, FSDP2-safe per-token logp), so CE = `-mean(logp)` over the response with prompt tokens masked.
- **SD3 task** (`unirl/models/sd3/sft_task.py`): reuses `SD3Bundle` / `SD3TextEmbedStage` / `SD3DiffusionStep.predict_noise` / `get_sigma_schedule`; adds the VAE image→latent encode. Target `noise - x0` matches `FlowSDEStrategy.step`'s velocity convention.
- **Recipes** `examples/sft/{qwen3,sd3}_sft_smoke.yaml` + `examples/sft/README.md`.

## Relationship to #181

This is the **generic SFT domain split out of #181** so it can land independently of Cosmos3. #181 adds the same skeleton *plus* a Cosmos3 (video/action BC) task; here the skeleton is decoupled and validated on two in-tree models instead. **Cosmos3 is NOT included** — it stays in #181, which can rebase onto this skeleton. No overlap in model code.

## Test Plan

- **CPU unit tests — 11 passed** (run under the unirl venv, no GPU / no real weights):
  - `tests/models/test_qwen3_sft_ce.py` — CE = `-mean(logp)`, prompt/response tokenization, `prompt_len`/`response_tokens` plumbing, empty-input guards.
  - `tests/models/test_sd3_sft_flow.py` — velocity target is `noise - x0` (perfect prediction → ~0 loss), interpolation `x_t=(1-σ)x0+σ·noise`, MSE against target, `guidance_scale=1.0`.
  - `tests/train/test_sft_data.py` — JSONL data source (from skeleton).
- **Import + contract smoke**: both tasks import and are `SFTTaskBase` subclasses with correct `block_class_names` (`Qwen3DecoderLayer` / `JointTransformerBlock`).
- `ruff check` over all new files — clean.
- **GPU e2e: NOT yet run** (draft). Recommended:
  - `python -m unirl.train_sft --config-name sft/qwen3_sft_smoke num_rollouts=6 eval_interval=3` → CE loss decreases, no NaN.
  - `python -m unirl.train_sft --config-name sft/sd3_sft_smoke num_rollouts=6 eval_interval=3` → MSE loss decreases, eval emits an image.
  - Success = both losses fall, demonstrating one skeleton fits AR (CE) and diffusion (MSE).

## Compatibility / Risk

- **Purely additive** (14 new files; zero edits to existing modules). GRPO/ReFL/rollout paths untouched.
- Recipes use LoRA to keep the smoke light; drop `lora_cfg` for full finetuning.
- Two spots flagged for GPU verification: (a) Qwen3 `_replay_aware_forward` install ordering vs FSDP2 wrap + no-NaN under bf16 (it disables cuDNN-SDP internally, needs sdpa/flash attn); (b) SD3 VAE encode dtype/scaling round-trip.

## Checklist

- [x] Reviewed the changed code and removed unrelated/generated artifacts.
- [x] Updated tests, docs, and configs where needed, or explained why not.

## EXP

HF SFT TRAINER QWEN3-4b base
截屏2026-07-10 03 07 52

unirl SFT TRAINER QWEN3-4b base
截屏2026-07-10 03 08 20

HF SFT TRAINER and unirl SFT TRAINER SD3
截屏2026-07-10 13 39 36

### CODE REVIEW

**Comment by leviking98z-rgb** [comment]
## Smoke + reference-parity results (1×8 H20, unirl venv)

Validated both tasks against **reference implementations** (stronger than eyeballing two noisy training curves — the loss *definitions* are shown to coincide numerically), plus a real training run for the AR side.

### Qwen3 (AR / cross-entropy)

**Numerical parity vs HF/TRL-standard CE** — same Qwen3-4B-Base, same batch, fp32, single GPU (`tests/_parity_qwen3_ce.py`): our `compute_loss` (`-mean(logp)` via `_replay_aware_forward`) vs the textbook masked shift-label `F.cross_entropy` (prompt = `-100`) that HF `Trainer` / TRL `SFTTrainer` compute:

```
[0] ref=1.18262  ours=1.18262  |Δ|=0.00e+00
[6] ref=3.15694  ours=3.15694  |Δ|=4.77e-07
max |Δ| = 4.77e-07   PARITY: PASS ✅
```

**Real training run** — Qwen3-4B-Base + Alpaca subset (512 instruction/response, no-`input` rows), LoRA, `num_rollouts=8`:

```
step 1/8  loss=1.83850  grad_norm=2.39
step 3/8  loss=1.18511  grad_norm=1.45
step 6/8  loss=0.95415  grad_norm=0.65
step 8/8  loss=1.37204  grad_norm=1.31
eval samples @ step 4, 8 written; clean exit, no NaN
```

CE loss descends, grad_norm healthy, eval sampling works.

### SD3 (diffusion / flow-matching MSE)

**Numerical parity vs diffusers-standard flow-matching** (`tests/_parity_sd3_flow.py`, CPU, random latents): our loss vs the textbook MSE against velocity target `v = noise - x0`:

```
exact-target loss max = 6.75e-15   (predict == target -> ~0)
vs textbook MSE(v=noise-x0) max |Δ| = 0.00e+00
SD3 FLOW PARITY: PASS ✅
```

Confirms the velocity-target convention (`noise - x0`, not `noise`) and the `x_t=(1-σ)x0+σ·noise` interpolation. A full SD3 GPU training run is pending a real image-caption dataset (the cluster's proxy blocks HF file-CDN downloads; metadata/API works but large blobs don't) — the math is proven correct meanwhile.

### Bug fixed during the smoke

`Qwen3SFTTask.load_record`: on **transformers 5.x**, `apply_chat_template(tokenize=True)` returns a `BatchEncoding` dict (iterating yields string keys → `too many dimensions 'str'`). Fixed with `return_dict=False`. Pushed in `4614a0f`.

**Comment by leviking98z-rgb** [comment]
## True parity vs `trl.SFTTrainer`'s actual loss

Followup to the parity note above: the earlier qwen3 check compared against a *hand-written* masked shift-label CE. Added the stronger check (`tests/_parity_qwen3_hf_native.py`, pushed `702889c`) — compare our `compute_loss` against the model's **own** CausalLM loss via `model(input_ids, labels=[-100]*prompt + response)`, which is exactly what `trl.SFTTrainer` / HF `Trainer` optimize. No hand-written reference can hide a masking/shift bug.

Two separate Qwen3-4B instances (stock `labels=` forward vs our `_replay_aware_forward`) rule out forward shadowing; both fed identical tokenization via `Qwen3SFTTask.load_record`. fp32, single GPU:

```
[0] hf_native=1.18262  ours=1.18262  |Δ|=0.00e+00
[6] hf_native=3.15694  ours=3.15694  |Δ|=4.77e-07
max |Δ| = 4.77e-07   TRUE PARITY vs HF/TRL loss: PASS ✅
```

Our SFT CE is numerically identical to the loss `SFTTrainer` optimizes.

**Comment by leviking98z-rgb** [comment]
## Longer runs — loss descent (qwen3) + eval images (sd3)

Followup on "is the loss just flat because too few steps?" — **yes for qwen3**; and for sd3, **flat MSE is expected — the eval images are what show learning**.

### Qwen3 CE — it descends with enough steps

The earlier 40/400-step runs looked flat. With lr 2e-4 and 1500 steps (~3 epochs of the 4k Alpaca subset), the CE clearly and monotonically drops:

```
first 30 steps:  1.222
step 301-400:    1.194
step 701-800:    1.061
step 1101-1200:  0.828
last 100 steps:  0.800
```

**1.22 → 0.80.** The earlier flatness was step-count + lr, not a bug (and the TRL reference was equally flat at 40 steps, confirming it's shared behavior, not our implementation). wandb run: `qwen3_unirl_lr2e4`.

### SD3 — flat MSE is normal; eval images show the style transfer

Diffusion training loss stays ~0.25 throughout (each step samples a random σ + noise → high-variance expectation over all noise levels; it does NOT trend down even when the model is learning). The right signal is the **eval samples**: after training on `lambda/naruto-blip-captions`, the SD3.5 LoRA generates clean anime-style portraits in the naruto aesthetic. Logged to wandb run `sd3_eval_images` (4 samples at steps 300/600).

### Summary of what these runs establish

- **Numerical parity** (the rigorous alignment): our CE == HF-native `model(labels=).loss` to |Δ|=4.8e-7; our flow-matching loss == diffusers `MSE(v=noise-x0)` to |Δ|=0.
- **Training works end-to-end**: qwen3 CE descends 1.22→0.80; sd3 produces on-style eval images.
- **UniRL vs reference curves overlay** in the same wandb project `unirl-sft-4curves` (qwen3: UniRL vs TRL SFTTrainer; sd3: UniRL vs diffusers).

All wandb runs: https://wandb.ai/leviking98z-zhejiang-university/unirl-sft-4curves
